zkml/benches/prove-model.rs

```diff
@@ -0,0 +1,112 @@
+
+fn new_transcript() -> Transcript {
+    #[cfg(feature = "blake")]
+    {
+        use transcript::blake::BlakeTranscript;
+        BlakeTranscript::new(b"bench")
+    }
+    #[cfg(not(feature = "blake"))]
+    {
+        default_transcript()
+    }
+}
+
+fn parse_model(model_data: &[u8]) -> anyhow::Result<(Model<Element>, ModelMetadata)> {
+    FloatOnnxLoader::from_bytes_with_scaling_strategy(model_data, AbsoluteMax::new())
+        .with_keep_float(true)
+        .build()
+}
+
+fn run_model<T: std::io::Read>(model_data: &[u8], inputs: T) {
+    let run_inputs = Input::from_reader(inputs).expect("failed to load inputs");
+    let (model, md) = parse_model(model_data).expect("failed to parse model");
+    let inputs = run_inputs.to_elements(&md);
+
+    let ctx = Some(
+        Context::<F, Pcs<F>>::generate(&model, None, None).expect("unable to generate context"),
+    );
+
+    for (i, input) in inputs.into_iter().enumerate() {
+        let input_tensor = model
+            .load_input_flat(vec![input])
+            .expect("failed to call load_input_flat on the model");
+
+        let trace = model
+            .run(&input_tensor)
+            .expect(&format!("input #{i} failed"));
+
+        let mut prover_transcript = new_transcript();
+        let prover = Prover::<_, _, _>::new(ctx.as_ref().unwrap(), &mut prover_transcript);
+        let proof = prover.prove(&trace).expect("unable to generate proof");
+
+        let mut verifier_transcript = new_transcript();
+        verify::<_, _, _>(
+            ctx.as_ref().unwrap().clone(),
+            proof,
+            trace.to_verifier_io(),
+            &mut verifier_transcript,
+        )
+        .expect("invalid proof");
+    }
+}
+
+fn criterion_benchmark(c: &mut Criterion) {
+    let mut group = c.benchmark_group("run-models");
+    group
+        .sample_size(20)
+        .measurement_time(std::time::Duration::from_secs(17));
+
+    group.bench_function("mlp", |b| {
+        b.iter(|| {
+            run_model(
+                include_bytes!("ref-files/mlp/model.onnx"),
+                zstd::Decoder::new(&include_bytes!("ref-files/mlp/input.json.zst")[..])
+                    .expect("failed to parse zstd"),
+            )
+        })
+    });
+    group.bench_function("cnn", |b| {
+        b.iter(|| {
+            run_model(
+                include_bytes!("ref-files/cnn/model.onnx"),
+                zstd::Decoder::new(&include_bytes!("ref-files/cnn/input.json.zst")[..])
+                    .expect("failed to parse zstd"),
+            )
+        })
+    });
+    group.bench_function("covid", |b| {
+        b.iter(|| {
+            run_model(
+                include_bytes!("ref-files/covid/model.onnx"),
+                zstd::Decoder::new(&include_bytes!("ref-files/covid/input.json.zst")[..])
+                    .expect("failed to parse zstd"),
+            )
+        })
+    });
+
+    group.finish();
+}
+
+criterion_group!(benches, criterion_benchmark);
```

Comment:
We should probably put options to run it like only 3 times or so no ? These bench are made for small functions usually no ? I'm a bit afraid of the CI time here..
Because the goal is to put that on CI right ? 
Otherwise, if that's *not* the goal, then i'm failing to see what's the point of this, vs using  the existing `bench.rs` that gives out much more datapoints.

Reply:
That's correct – the numeric parameters are not definitive yet, I'm currently running some experiments to see what would be a good number before actually enabling it on CI.